go/tools/gopackagesdriver/pkgjson/main.go

```diff
@@ -0,0 +1,228 @@
+	if strings.HasPrefix(p, "__BAZEL_OUTPUT_BASE__") {
+		return strings.TrimPrefix(p, "__BAZEL_OUTPUT_BASE__/")
+	}
+	if strings.HasPrefix(p, "__BAZEL_EXECROOT__") {
+		return strings.TrimPrefix(p, "__BAZEL_EXECROOT__/")
+	}
+	if strings.HasPrefix(p, "__BAZEL_WORKSPACE__") {
+		return strings.TrimPrefix(p, "__BAZEL_WORKSPACE__/")
+	}
+	return p
+}
+
+// processCgoFiles ensures that the CompiledGoFiles field of the package json does not contain
+// any cgo source files and contains cgo generated files.
+func processCgoFiles(cgoOutDir string, p *pkgJson, pathResolver func(string) string) error {
+	cf, err := filterCgoSourceFiles(p.CompiledGoFiles, pathResolver)
+	if err != nil {
+		return err
+	}
+	p.CompiledGoFiles = cf.goSourceFiles
+	if cgoOutDir == "" {
+		return nil
+	}
+	cgoSrcs, err := cgoGoSrcs(cgoOutDir, cf.cgoSourceFiles, pathResolver)
+	if err != nil {
+		return err
+	}
+	p.CompiledGoFiles = append(p.CompiledGoFiles, cgoSrcs...)
+	return nil
+}
+
+type cgoFiles struct {
+	cgoSourceFiles []string
+	goSourceFiles []string
+}
+
+func filterCgoSourceFiles(compiledGoFiles []string, resolvePath func(string) string) (cgoFiles, error) {
+	filtered := make([]string, 0, len(compiledGoFiles))
+	var cgoSourceFiles []string
+	fset := token.NewFileSet()
+	for _, file := range compiledGoFiles {
+		// Note: We check resolved file paths to get a path relative to the execroot
+		// but we add the unresolved file path to the package json as the gopackagesdriver
+		// uses the special bazel prefixes to resolve to absolute paths.
+		resolvedPath := resolvePath(file)
+		f, err := parser.ParseFile(fset, resolvedPath, nil, parser.ImportsOnly)
+		if err != nil {
+			return cgoFiles{}, err
+		}
+		var skip bool
+		for _, rawImport := range f.Imports {
+			imp, err := strconv.Unquote(rawImport.Path.Value)
+			if err != nil {
+				continue
+			}
+			if imp == "C" {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			// Skip Cgo preprocessed files.
+			cgoSourceFiles = append(cgoSourceFiles, file)
+		} else {
+			filtered = append(filtered, file)
+		}
+	}
+	return cgoFiles{
+		cgoSourceFiles: cgoSourceFiles,
+		goSourceFiles:  filtered,
+	}, nil
+}
+
+func cgoGoSrcs(cgoGeneratedDir string, cgoSourceFiles []string, resolvePath func(string) string) ([]string, error) {
+	var cgoGeneratedFiles []string
+	// Only include cgo generated files if they were generated by Bazel, using
+	// the _cgo_gotypes.go file name as a marker for cgo files as it's always
+	// present if the package contains cgo generated code.
+	//
+	// Note: We check resolved file paths to get a path relative to the execroot
+	// but we add the unresolved file path to the package json as the gopackagesdriver
+	// uses the special bazel prefixes to resolve to absolute paths.
+	resolvedCgoGeneratedDir := resolvePath(cgoGeneratedDir)
+	path := filepath.Join(resolvedCgoGeneratedDir, "_cgo_gotypes.go")
+	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
+		return nil, nil
+	}
```

Comment:
```suggestion
	cgotypes_path := filepath.Join(resolvedCgoGeneratedDir, "_cgo_gotypes.go")
	if _, err := os.Stat(cgotypes_path); err != nil && os.IsNotExist(err) {
		return nil, nil
	}
```
nit: `path` may be confused by the path import by readers.